R/qra_fit_convenience.R

```diff
@@ -1104,47 +1120,83 @@ impute_missing_per_quantile <- function(qfm, impute_method = 'mean') {
 
   X_na <- is.na(qfm)
 
-  missingness_groups <- X_na %>%
+  missingness_patterns <- X_na %>%
     as.data.frame() %>%
     mutate(row_num = dplyr::row_number()) %>%
-    dplyr::group_by_at(seq_len(ncol(.) - 1)) %>%
-    dplyr::summarise(row_inds = list(row_num))
+    dplyr::group_by(across(seq_len(ncol(.) - 1))) %>%
+    dplyr::summarise(row_inds = list(row_num), .groups = "drop")
 
   qfm_imputed <- qfm
   qfm_imputed[is.na(qfm_imputed)] <- 0.0
 
   weight_transfer <- matrix(0, nrow = num_models, ncol = num_models)
 
-  for(i in seq_len(nrow(missingness_groups))) {
-    row_inds <- missingness_groups$row_inds[[i]]
+  if (weight_transfer_per_group) {
+    row_groups <- attr(qfm, 'row_index') %>% 
+      mutate(row_num = dplyr::row_number()) %>%
+      dplyr::group_by(!!!syms(weight_transfer_group_factors)) %>% 
+      dplyr::summarise(row_inds_per_group = list(row_num), .groups = "drop") %>% 
+      mutate(weight_transfer = list(weight_transfer))
+  }
+
+  for(i in seq_len(nrow(missingness_patterns))) {
+    row_inds <- missingness_patterns$row_inds[[i]]
 
+    # intialize as identity
     impute_mat <- diag(num_models)
 
     col_inds <- which(quantile_levels == unique_quantile_levels[1])
     temp <- !is.na(unclass(qfm)[row_inds[1], col_inds])
     temp <- temp / sum(temp)
 
+    # form transfer matrix based on first quantile for first 
+    # row (location-date-target) with missingness pattern
     for(j_ind in seq_along(col_inds)) {
       j <- col_inds[j_ind]
+      # replace e_i with 'average of non-missing' column
       if(is.na(qfm[row_inds[1], j])) {
         impute_mat[, j_ind] <- temp
       }
     }
 
+    # use this matrix to impute all quantiles in all rows with i'th miss pattern
     for(quantile_level in unique_quantile_levels) {
       col_inds <- which(quantile_levels == quantile_level)
       qfm_imputed[row_inds, col_inds] <-
         qfm_imputed[row_inds, col_inds, drop = FALSE] %*% impute_mat
     }
 
     weight_transfer <- weight_transfer + length(row_inds) * impute_mat
+    if (weight_transfer_per_group) {
+      row_groups <- row_groups %>% mutate(
+        weight_transfer = purrr::map2(
+          weight_transfer, row_inds_per_group,
+          ~ .x + length(intersect(row_inds, .y)) * impute_mat 
+          )
+        )
```

Comment:
In this area, suppose we have `weight_transfer_per_group == TRUE`, and there are two `row_groups`, e.g. for two different locations, and both groups had some observations with the missingness pattern we're looking at for index `i` -- say, 1 observation for group 1 and 2 observations for group 2. It seems like on the new line 1169, we would add 3 copies of `impute_mat` to `weight_transfer`, and then on line 1174 we would add another 1 copy for group 1 and another 2 copies for group 2 for the location-specific weight transfer matrices.  Basically, I would have thought we'd only want to do the operation on line 1169 in an else clause here, if `weight_transfer_per_group` was `FALSE`.  Am I missing something here though?

Reply:
I probably should use more/different names here but I don't think there is a problem in the logic because the weight_transfer on 1169 is a different object than the weight_transfer on 1174 (a column of row_groups)... and only one of them end up getting returned depending on `weight_transfer_per_group`.

Reply:
that makes sense, thanks